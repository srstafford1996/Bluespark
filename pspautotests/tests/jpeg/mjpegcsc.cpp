#include "shared.h"
#include <malloc.h>
#include <psputility.h>
#include <psputils.h>

extern "C" int sceJpegMJpegCsc(void *dest, const void *srcYCbCr, uint32_t widthHeight, int stride);

static uint8_t ycbcr[1572864];
static uint32_t *frame;
static const uint32_t FRAME_SIZE = 1024 * 1024 * 4;

static void testJpegCsc(const char *title, bool useDest, bool useSrc, uint32_t widthHeight, int stride) {
	memset(frame, 0xCC, FRAME_SIZE + 4);
	sceKernelDcacheWritebackInvalidateRange(frame, FRAME_SIZE + 4);

	int result = sceJpegMJpegCsc(useDest ? frame : NULL, useSrc ? ycbcr : NULL, widthHeight, stride);
	size_t len = 0;
	for (int i = 1024 * 1024; i >= 0; --i) {
		if (frame[i] != 0xCCCCCCCC) {
			len = (i + 1) * sizeof(uint32_t);
			break;
		}
	}
	size_t rowLen = 0;
	for (int i = 0; i < (int)len; ++i) {
		if (frame[i] == 0xCCCCCCCC) {
			rowLen = i * sizeof(uint32_t);
			break;
		}
	}
	checkpoint("%s: %08x (wrote %d bytes, %d per row)", title, result, len, rowLen);
	if (!useSrc) {
		checkpoint("%s: wrote %08x", title, frame[0]);
	}
}

static void testJpegCscPattern(const char *title, int y, int cb, int cr) {
	sceKernelDcacheWritebackInvalidateRange(frame, FRAME_SIZE);
	memset(frame, 0xCC, FRAME_SIZE);
	sceKernelDcacheWritebackInvalidateRange(frame, FRAME_SIZE);

	int w = 480;
	int h = 272;
	int ysize = w * h;
	int cbcrSize = ysize >> 2;

	sceKernelDcacheWritebackInvalidateRange(ycbcr, sizeof(ycbcr));
	memset(ycbcr, 0, sizeof(ycbcr));
	memset(ycbcr, 0xFF, ysize - 256 + y);
	memset(ycbcr + ysize, 0x80, cbcrSize - 64 + cb);
	memset(ycbcr + ysize + cbcrSize, 0x80, cbcrSize - 64 + cr);
	sceKernelDcacheWritebackInvalidateRange(ycbcr, sizeof(ycbcr));

	sceKernelDcacheWritebackInvalidateAll();
	int result = sceJpegMJpegCsc(frame, ycbcr, (w << 16) | h, w);
	sceKernelDcacheWritebackInvalidateAll();
	checkpoint(NULL);
	schedf("%s: %08x =", title, result);
	uint32_t last = frame[0];
	int lasti = 0;
	for (int i = 1; i < w * h; ++i) {
		if (frame[i] != last) {
			schedf(" %d-%d %08x", lasti, i - 1, last);
			last = frame[i];
			lasti = i;
		}
	}
	schedf(" %d-%d %08x\n", lasti, w * h - 1, last);
}

static void testJpegCscValue(const char *title, uint8_t yy, uint8_t cb, uint8_t cr) {
	sceKernelDcacheWritebackInvalidateRange(frame, 4);
	memset(frame, 0xCC, 4);
	sceKernelDcacheWritebackInvalidateRange(frame, 4);

	sceKernelDcacheWritebackInvalidateRange(ycbcr, sizeof(ycbcr));
	memset(ycbcr, 0, 768);
	ycbcr[0] = yy;
	ycbcr[256] = cb;
	ycbcr[256 + 64] = cr;
	sceKernelDcacheWritebackInvalidateRange(ycbcr, sizeof(ycbcr));

	int result = sceJpegMJpegCsc(frame, ycbcr, 0x00100010, 0x0010);
	checkpoint("%s: %08x (%08x)", title, result, frame[0]);
}

extern "C" int main(int argc, char *argv[]) {
	sceUtilityLoadModule(PSP_MODULE_AV_AVCODEC);

	frame = (uint32_t *)malloc(FRAME_SIZE + 4);

	checkpointNext("Basic:");
	testJpegCsc("  Before init", true, true, 0x00100010, 0x0010);
	checkpoint("  Init: %08x", sceJpegInitMJpeg());
	testJpegCsc("  After init", true, true, 0x00100010, 0x0010);
	checkpoint("  Finish: %08x", sceJpegFinishMJpeg());
	testJpegCsc("  After finish", true, true, 0x00100010, 0x0010);
	sceJpegInitMJpeg();

	checkpointNext("Buffers:");
	testJpegCsc("  Output NULL", false, true, 0x00100010, 0x0010);
	testJpegCsc("  Input NULL", true, false, 0x00100010, 0x0010);

	checkpointNext("Sizes:");
	testJpegCsc("  Wider than stride", true, true, 0x00200020, 0x0010);
	testJpegCsc("  Thinner than stride", true, true, 0x00100020, 0x0020);
	testJpegCsc("  Odd size", true, true, 0x05550555, 0x0010);
	testJpegCsc("  Large width", true, true, 0xFFFF0001, 0x0010);
	testJpegCsc("  721 width", true, true, (721 << 16) | 1, 0x0010);
	testJpegCsc("  720 width", true, true, (720 << 16) | 1, 0x0010);
	testJpegCsc("  719 width", true, true, (719 << 16) | 1, 0x0010);
	testJpegCsc("  720x32", true, true, (720 << 16) | 32, 0x0010);
	testJpegCsc("  720x16", true, true, (720 << 16) | 16, 0x0010);
	testJpegCsc("  720x8", true, true, (720 << 16) | 8, 0x0010);
	testJpegCsc("  Zero width", true, true, 0x00000010, 0x0010);
	testJpegCsc("  Small width", true, true, 0x00070010, 0x0010);
	testJpegCsc("  Large height", true, true, 0x0008FFFF, 0x0008);
	testJpegCsc("  481 height", true, true, 0x00080000 | 481, 0x0008);
	testJpegCsc("  480 height", true, true, 0x00080000 | 480, 0x0008);
	testJpegCsc("  479 height", true, true, 0x00080000 | 479, 0x0008);
	testJpegCsc("  Zero height", true, true, 0x00100000, 0x0010);

	checkpointNext("Strides:");
	// Hangs.
	//testJpegCsc("  Negative stride", true, true, 0x00100010, -1);
	testJpegCsc("  Negative stride", true, true, 0x00100010, -4);
	testJpegCsc("  Zero stride", true, true, 0x00100010, 0);
	testJpegCsc("  One stride", true, true, 0x00100010, 1);
	testJpegCsc("  1024 stride", true, true, 0x00100010, 1024);
	testJpegCsc("  1025 stride", true, true, 0x00100010, 1025);
	testJpegCsc("  Large stride", true, true, 0x00100002, 0x7FFFFFFF);
	testJpegCsc("  Very negative stride", true, true, 0x00100002, 0x80000000);

	checkpointNext("X/Y Pattern:");
	testJpegCscPattern("  2x2 full", 256, 64, 64);
	testJpegCscPattern("  2x2 short y", 255, 64, 64);
	testJpegCscPattern("  2x2 short cb", 256, 63, 64);
	testJpegCscPattern("  2x2 short cr", 256, 64, 63);

	checkpointNext("Bits:");
	testJpegCscValue("  2x2 Black", 0, 128, 128);
	testJpegCscValue("  2x2 White", 255, 128, 128);
	testJpegCscValue("  2x2 0 Cb", 255, 0, 128);
	testJpegCscValue("  2x2 1 Cb", 255, 1, 128);
	testJpegCscValue("  2x2 254 Cb", 255, 254, 128);
	testJpegCscValue("  2x2 255 Cb", 255, 255, 128);
	testJpegCscValue("  2x2 0 Cr", 255, 128, 0);
	testJpegCscValue("  2x2 1 Cr", 255, 128, 1);
	testJpegCscValue("  2x2 254 Cr", 255, 128, 254);
	testJpegCscValue("  2x2 255 Cr", 255, 128, 255);

	free(frame);
	sceUtilityUnloadModule(PSP_MODULE_AV_AVCODEC);
	return 0;
}