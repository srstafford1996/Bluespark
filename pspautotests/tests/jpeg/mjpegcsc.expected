[x] Basic:
[x]   Before init: 80000001 (wrote 0 bytes, 0 per row)
[r]   Init: 00000000
[r]   After init: 00000000 (wrote 1024 bytes, 1024 per row)
[x]   Finish: 00000000
[x]   After finish: 80000001 (wrote 0 bytes, 0 per row)

[x] Buffers:
[x]   Output NULL: 80000103 (wrote 0 bytes, 0 per row)
[r]   Input NULL: 00000000 (wrote 1024 bytes, 1024 per row)
[x]   Input NULL: wrote 00008600

[x] Sizes:
[r]   Wider than stride: 00000000 (wrote 2112 bytes, 2112 per row)
[r]   Thinner than stride: 00000000 (wrote 4032 bytes, 64 per row)
[x]   Odd size: 80650020 (wrote 0 bytes, 0 per row)
[x]   Large width: 80650020 (wrote 0 bytes, 0 per row)
[x]   721 width: 80650020 (wrote 0 bytes, 0 per row)
[x]   720 width: 00000000 (wrote 160 bytes, 0 per row)
[x]   719 width: 00000000 (wrote 160 bytes, 0 per row)
[r]   720x32: 00000000 (wrote 4864 bytes, 4864 per row)
[r]   720x16: 00000000 (wrote 3840 bytes, 3840 per row)
[x]   720x8: 00000000 (wrote 384 bytes, 0 per row)
[r]   Zero width: 00000000 (wrote 5056 bytes, 5056 per row)
[r]   Small width: 00000000 (wrote 5056 bytes, 5056 per row)
[x]   Large height: 80650020 (wrote 0 bytes, 0 per row)
[x]   481 height: 80650020 (wrote 0 bytes, 0 per row)
[r]   480 height: 00000000 (wrote 19424 bytes, 19424 per row)
[r]   479 height: 00000000 (wrote 19152 bytes, 19152 per row)
[x]   Zero height: 00000000 (wrote 64 bytes, 64 per row)

[x] Strides:
[r]   Negative stride: 00000000 (wrote 57296 bytes, 64 per row)
[r]   Zero stride: 00000000 (wrote 64 bytes, 64 per row)
[r]   One stride: 00000000 (wrote 120 bytes, 120 per row)
[r]   1024 stride: 00000000 (wrote 61504 bytes, 64 per row)
[x]   1025 stride: 80650020 (wrote 0 bytes, 0 per row)
[x]   Large stride: 80650020 (wrote 0 bytes, 0 per row)
[x]   Very negative stride: 00000000 (wrote 64 bytes, 64 per row)

[x] X/Y Pattern:
[r]   2x2 full: 00000000 = 0-3 00ffffff 4-4 00505050 5-5 00575757 6-6 009b9b9b 7-7 00080808 8-15 00ffd3ff 16-487 00ffffff 488-495 00ffd3ff 496-130559 00ffffff
[r]   2x2 short y: 00000000 = 0-3 00ffffff 4-4 00505050 5-5 00575757 6-6 009b9b9b 7-7 00080808 8-13 00ffd3ff 14-15 001effff 16-487 00ffffff 488-493 00ffd3ff 494-495 001effff 496-130559 00ffffff
[r]   2x2 short cb: 00000000 = 0-3 00ffffff 4-4 00505050 5-5 00575757 6-6 009b9b9b 7-7 00080808 8-13 00ffd3ff 14-15 00ffff4d 16-487 00ffffff 488-493 00ffd3ff 494-495 00ffff4d 496-130559 00ffffff
[r]   2x2 short cr: 00000000 = 0-3 00ffffff 4-4 00505050 5-5 00575757 6-6 009b9b9b 7-7 00080808 8-15 00ffd3ff 16-487 00ffffff 488-495 00ffd3ff 496-130559 00ffffff

[x] Bits:
[r]   2x2 Black: 00000000 (00000000)
[r]   2x2 White: 00000000 (00ffffff)
[r]   2x2 0 Cb: 00000000 (001effff)
[r]   2x2 1 Cb: 00000000 (001effff)
[r]   2x2 254 Cb: 00000000 (00ffd4ff)
[r]   2x2 255 Cb: 00000000 (00ffd3ff)
[r]   2x2 0 Cr: 00000000 (00ffff4d)
[r]   2x2 1 Cr: 00000000 (00ffff4d)
[r]   2x2 254 Cr: 00000000 (00ffa5ff)
[r]   2x2 255 Cr: 00000000 (00ffa5ff)
